Implement "fallback" transformation when selected operation fails

nyalldawson opened the pull request:
When the selected coordinate operation fails to project coordinates, try falling back to a default PROJ-provided operation to transform the coordinates

This allows **some** transformation of coordinates to occur in the situation:
- user has selected a coordinate operation utilising a grid shift file
- one or more points which fall OUTSIDE the grid file are transformed

Not for merge -- likely needs some form of user-visible-warning when this occurs

Refs #33929

nyalldawson committed:
When the selected coordinate operation fails to project coordinates,
try falling back to a default PROJ-provided operation to transform the
coordinates

This allows **some** transformation of coordinates to occur in the situation:
- user has selected a coordinate operation utilising a grid shift file
- one or more points which fall OUTSIDE the grid file are transformed

Not for merge -- likely needs some form of user-visible-warning when
this occurs

Refs #33929

nyalldawson commented:
@rouault keen to hear your thoughts on this!

rouault commented on src/core/qgscoordinatetransform.cpp:
```diff
@@ -637,6 +637,13 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
     return;
   }
 
+  double *xprev = new double[ numPoints ];
```
should be restricted #if PROJ_VERSION_MAJOR>=6 to avoid mem leak

rouault commented on src/core/qgscoordinatetransform.cpp:
```diff
@@ -637,6 +637,13 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
+  double *yprev = new double[ numPoints ];
+  memcpy( yprev, y, sizeof( double ) * numPoints );
+  double *zprev = new double[ numPoints ];
+  memcpy( zprev, z, sizeof( double ) * numPoints );
```
this method cannot be called with z==NULL right ?

nyalldawson replied:
right -- all code paths leading here pad out the z array with 0's if it doesn't already exist

rouault commented on src/core/qgscoordinatetransform.cpp:
```diff
@@ -707,6 +720,45 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
 
+#if PROJ_VERSION_MAJOR>=6
+
+  if ( actualRes != 0 )
```
actualRes will not be reliable for numPoints > 1. you'd probably have to iterate over all transformed points to check their x != HUGE_VAL

nyalldawson replied:
Why is that? Does proj_errno get reset mid-operation when transforming multiple points?

rouault replied:
I'm not sure. Error handling in PROJ is super confusing (to me at least)... See https://github.com/OSGeo/PROJ/issues/1856

nyalldawson committed:
Show warnings on fallback transform

nyalldawson committed:
We can't always rely on proj_errno to indicate multi-point transform
failures, so manually scan for inf values to detect failures which
should result in the fallback transform usage

nyalldawson committed:
Add mechanism to disable user-visible warnings on use of fallback transforms
for a particular QgsCoordinateTransform object

This can be set for transforms where we definitively know that a "ballpark"
result is acceptable, e.g. when transforming global layer extents to
a localized area-of-use CRS

Refs #33929

nyalldawson committed:
Mark some transforms as approximate

nyalldawson committed:
Remove "used" operation from handler

We can't populate this, because it's not a singular proj operation

nyalldawson committed:
Don't try to get the operation details for the instantiated operation,
instead correctly use the desired operation for the handler

nyalldawson committed:
Don't include proj string in QgsCsException messages

This is too costly to calculate, and results in a lot of processing
time when raster rendering transform errors occur (since they occur
on a one-by-one basis and in large quantities)

nyalldawson committed:
When using the fallback transform, we can't trust the results of proj_errno

It's giving us incorrect error numbers (such as "failed to load datum shift file",
which is definitely incorrect for a default proj created operation!)
so we resort to testing values ourselves...

nyalldawson committed:
Only warn a single time per project for fallback transform usage

nyalldawson committed:
Much much faster QgsCoordinateReferenceSystem equality operator in many cases

nyalldawson committed:
ifdef out noisy debug code

nyalldawson committed:
Friendlier warning message when fallback transform is used

Hopefully giving enough clues for users to diagnose the error

nyalldawson committed:
Add API for disabling fallback handler and for determining after a transform whether a fallback operation occurred

nyalldawson committed:
[processing] Log use of ballpark transforms to processing log instead
of to the main window message bar

More processing algorithms will require this, but it's a template for
the others to be adapted

nyalldawson commented:
Getting closer... here's how things look now:

![image](https://user-images.githubusercontent.com/1829991/74503452-a7332380-4f3c-11ea-83bd-36821eb19039.png)


![image](https://user-images.githubusercontent.com/1829991/74503458-ad290480-4f3c-11ea-8e05-da8b965d90e4.png)


@rouault I'm curious to hear your thoughts on https://github.com/qgis/QGIS/pull/34273/commits/3f2c6733337164bd4c838906b6cc3602e2c2d18b -- is this a proj bug?

nyalldawson commented:
Little screencast:

![Peek 2020-02-14 15-16](https://user-images.githubusercontent.com/1829991/74503572-0abd5100-4f3d-11ea-8c89-08397f371c6b.gif)

nyalldawson committed:
Ballpark warning should be warning level, not critical

nyalldawson committed:
Crappy US spelling

rouault commented:
> I'm curious to hear your thoughts on [3f2c673](https://github.com/qgis/QGIS/commit/3f2c6733337164bd4c838906b6cc3602e2c2d18b) -- is this a proj bug?

maybe... A reproducer ?

nirvn commented:
Tested the branch, I can confirm it fixes the linked issues. No regressions spotted so far either.

nirvn commented:
@nyalldawson , the plan is to introduce this in 3.12 and backport it in 3.10.4 (ie the point release after the one coming in a few days)?

nyalldawson commented:
@nirvn I'm unsure... Its an important fix, but relatively high risk...

nirvn commented:
For what it's worth, I would backport. I think the benefits outweight the risks. Plus, it'd be good for this LTR to be as sync'ed as possible on the projection/transformation front, it'll ease backports for the year to come.

nyalldawson committed:
Add API to QgsCoordinateTransformContext to prevent use of fallback
ballpark transforms for a source/dest CRS pair

nyalldawson committed:
Add API to QgsCoordinateTransform to control fallback availability
on a transform-by-transform basis, populate correctly from context

nyalldawson committed:
Fix allow fallback determination for reversed operations

nyalldawson committed:
Fix interaction of QgsCoordinateTransform::setBallparkTransformsAreAppropriate
and setAllowFallbackTransforms and improve dox

setBallparkTransformsAreAppropriate overrules the setAllowFallbackTransforms
setting

nyalldawson committed:
Less noisy transform failure errors

- Don't spew a whole lot of "rethrowing expection" log messages on the
default log level
- Don't output tons of duplicate transform error messages - instead
only output the message if its different to the last error encountered
(avoids flooding the console when transform failures occur during
raster projection, resulting in quasi-hangs)

nyalldawson committed:
[needs-docs] Add setting to control whether the coordinate operation
for transformation between a source and destination CRS pair should
permit falling back to a ballpark transform when required

This defaults to true (i.e. similar-ish behavior to proj 4), but allows
users who require a specific coordinate operation the ability to prevent
any fallback ballpark transformations occurring

nyalldawson committed:
Disable "Fallback" option when coordinate transforms are used in processing

nyalldawson committed:
Fix doxygen warning

nyalldawson commented:
Should be ready for merge now...

nyalldawson committed:
Fix build
